helix-view/src/editor.rs

```diff
@@ -1269,4 +1305,32 @@ impl Editor {
         .await
         .map(|_| ())
     }
+
+    pub async fn wait_event(&mut self) -> EditorEvent {
+        let mut saves: FuturesUnordered<_> = self
+            .documents
+            .values_mut()
+            .map(Document::await_save)
+            .collect();
```

Comment:
This seems expensive to do per event. Could we keep a `saves: FuturesUnordered` on the editor and push an `await_save` onto it when `editor.save` is called?

Reply:
That way you also don't need to store `current_save` on the document to prevent it from being dropped

Reply:
Yeah that's a good idea, will do

Reply:
To get this to work we need to add `editor.save`, but I think that's ok since we already do this for `open`/`close`/`switch`

Reply:
Just thought about this a bit, and I'm afraid it might become problematic. If we just chuck a future into an unordered pool like this every time we do a `:write`, then on quick successive saves of the same file, we could end up doing the saves out of order again. You could end up with this sequence:

1. `:write` 1 adds a save to the document's save queue, and adds the await operation to `editor.saves`
2. `:write` 2 adds a save to the document's save queue, and adds the await operation to `editor.saves`
3. event loop awaits, and happens to choose write 1's future to poll. This pulls write 1 off of the document's internal save queue
4. Part way through writing, the thread is yielded, other stuff happens
5. We go through another iteration of the main event loop. This time the select happens to choose doc 2's write future and polls. This causes the second write to be pulled from the document's internal save queue and begin writing
6. Write 2 finishes
7. Next event loop, write 1 finishes

I think it could basically take us back to square one. Although, maybe a regular `Vec<Future<_>>` could be used? And once the await finishes, pop it off? That way the futures always happen in order, and never get dropped.

Reply:
Well there's also a `FuturesOrdered` which acts as a queue: https://docs.rs/futures/latest/futures/stream/struct.FuturesOrdered.html

Reply:
Okay I understand this better now. We want a similar setup to the LSP queue: `SelectAll<UnboundedReceiver<_>>`. This way there's no need for `current_save` and `SelectAll` gets to own the `save_receiver` (it'll also get removed from the `SelectAll` if the document ever gets dropped and `save_sender` is closed).

To simplify further I'd also make `async fn save` on the doc return a simple save future and have the `save_sender`/queuing live in a separate `HashMap` on the editor. That way `editor.save` calls `doc.save`, then handles all the queuing and sequential writes.